File: 07_sam/01_api_basics/python/01_segment_image.py
#!/usr/bin/env python3
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

from __future__ import annotations

import argparse
import asyncio
from pathlib import Path

from sam_api import segment_image, write_outputs


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Segment concepts in one image.")
    parser.add_argument("image", type=Path)
    parser.add_argument(
        "--concept",
        action="append",
        required=True,
        help="short noun phrase; repeat to make a separate request",
    )
    parser.add_argument("--output", type=Path, required=True)
    return parser.parse_args()


async def main() -> None:
    args = parse_args()
    if not args.image.is_file():
        raise FileNotFoundError(args.image)
    results = []
    for concept in args.concept:
        print(f"Segmenting {concept!r}...")
        results.append((concept, await segment_image(args.image, concept)))
    write_outputs(args.output, results)


if __name__ == "__main__":
    asyncio.run(main())

File: 07_sam/01_api_basics/typescript/01-segment-image.ts
#!/usr/bin/env -S npx tsx
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { access } from 'node:fs/promises';

import { segmentImage, writeOutputs } from './sam-api.js';

function optionValues(name: string): string[] {
  const values: string[] = [];
  for (let index = 0; index < process.argv.length; index += 1) {
    if (process.argv[index] === name && index + 1 < process.argv.length) {
      values.push(process.argv[index + 1]!);
      index += 1;
    }
  }
  return values;
}

const filename = process.argv[2]?.startsWith('--') ? undefined : process.argv[2];
const concepts = optionValues('--concept');
const output = optionValues('--output').at(-1);
if (!filename || concepts.length === 0 || !output) {
  throw new Error(
    'Usage: 01-segment-image.ts <image> --concept <noun> [--concept <noun>] --output <dir>',
  );
}
await access(filename);

const results = [];
for (const concept of concepts) {
  console.log(`Segmenting ${JSON.stringify(concept)}...`);
  results.push([concept, await segmentImage(filename, concept)] as const);
}
await writeOutputs(output, results);
